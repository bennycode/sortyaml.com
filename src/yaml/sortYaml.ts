import {ParsedNode, parseDocument, YAMLMap, YAMLSeq} from 'yaml';

function sortDeep(node: ParsedNode | null): void {
  if (node instanceof YAMLMap) {
    node.items.sort((itemA, itemB) => (itemA.key < itemB.key ? -1 : itemA.key > itemB.key ? 1 : 0));
    node.items.forEach(item => sortDeep(item.value));
  } else if (node instanceof YAMLSeq) {
    node.items.forEach(item => sortDeep(item));
  }
}

export function sortYaml(text: string): string {
  try {
    const document = parseDocument(text);
    sortDeep(document.contents);
    return document.toString().trim();
  } catch {
    return `Invalid input`;
  }
}
