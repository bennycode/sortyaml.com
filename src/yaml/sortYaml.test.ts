import {sortYaml} from './sortYaml';
import fs from 'fs';
import path from "path";

describe('sortYaml', () => {
  it('sorts YAML content by key', async () => {
    const input = await fs.promises.readFile(path.join(__dirname, '../fixtures/unsorted.yml'), 'utf-8');
    const actual = sortYaml(input);
    expect(actual).toMatchSnapshot();
  });
});
