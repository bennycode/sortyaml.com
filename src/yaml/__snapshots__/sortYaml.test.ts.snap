// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortYaml sorts YAML content by key 1`] = `
"updates:
  - directory: '/'
    package-ecosystem: 'github-actions'
    schedule:
      day: 'tuesday'
      interval: 'monthly'
      time: '02:00'
      timezone: 'Europe/Berlin'
  - directory: '/'
    package-ecosystem: 'npm'
    schedule:
      day: 'tuesday'
      interval: 'monthly'
      time: '02:00'
      timezone: 'Europe/Berlin'
version: 2"
`;
